openedx/core/djangoapps/discussions/models.py

```diff
@@ -0,0 +1,113 @@
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+from jsonfield import JSONField
+from model_utils.models import TimeStampedModel
+from opaque_keys.edx.django.models import LearningContextKeyField
+from organizations.models import Organization
+from simple_history.models import HistoricalRecords
+
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+
+
+class DiscussionProviderConfig(TimeStampedModel):
+    """
+    Configuration model to store configuration for Discussions applications.
+    """
+
+    name = models.CharField(
+        blank=False,
+        max_length=100,
+        help_text=_("A user-friendly name for this configuration. e.g. SomeOrg Discourse")
+    )
+    provider = models.CharField(
+        blank=False,
+        db_index=True,
+        max_length=100,
+        verbose_name=_("Discussion provider"),
+        help_text=_("The discussion tool/provider."),
+    )
+    config = JSONField(
+        blank=False,
+        default={},
+        help_text=_("The configuration data for this provider."),
+    )
+    restrict_to_site = models.ForeignKey(
+        to=Site,
+        blank=True,
+        null=True,
+        on_delete=models.CASCADE,
+        help_text=_("Optionally restrict this config for use only on this site."),
+        db_index=True,
+    )
+    restrict_to_org = models.ForeignKey(
+        to=Organization,
+        blank=True,
+        null=True,
+        on_delete=models.CASCADE,
+        help_text=_("Optionally restrict this config for use only on this Organization."),
+        db_index=True,
+    )
+
+    history = HistoricalRecords()
+
+    def clean(self):
+        if self.restrict_to_org and self.restrict_to_site:
+            raise ValidationError("Can only set one form of restriction, site or org.")
+
+    def __str__(self):
+        return "{name} provider={provider} restricted to [site={site} org={org}]".format(
+            provider=self.provider,
+            name=self.name,
+            site=self.restrict_to_site,
+            org=self.restrict_to_org
+        )
+
+
+class LearningContextDiscussionConfig(TimeStampedModel):
+    """
+    Associates a learning context with a :class:`DiscussionProviderConfig`.
+
+    Also allows overriding some configuration on a course-by-course basis.
+    """
+
+    context_key = LearningContextKeyField(
+        primary_key=True,
+        db_index=True,
+        unique=True,
+        max_length=255,
+        verbose_name=_("Learning Context"),
+    )
+    enabled = models.BooleanField(
+        default=True,
+        help_text=_("If disabled, the discussions in the associated learning context/course will be disabled.")
+    )
+    provider_config = models.ForeignKey(
+        to=DiscussionProviderConfig,
+        null=True,
+        blank=True,
+        help_text=_("The configuration to use for this learning context."),
+        on_delete=models.SET_NULL,
+    )
+    config_overrides = JSONField(
```

Comment:
Can we talk a little about these `config_overrides` and how they relate to `config`?

i.e., can we give some examples of what will likely be in `config`, and what likely goes in `config_overrides`?  

I'm trying to understand how common it'll be to override config at the learning context level, and whether we're likely to override just _part_ of it, whether most/all of it would commonly be overridden, or whether it's possible to make a non-sensical configuration by overriding only a part of a set of fields, such as overriding a URL/tool identifier but not a secret key, for instance, making a configuration with the wrong key for the service.

Do we have known use cases for needing the overrides?  Or are we building in flexibility in "just in case"?  (I'm assuming the former, but don't have insight into the particular fields that would go into these configs...)

Reply:
To summarize the above:

1. I'd love to understand the detailed fields we expect to go into `config`
2. And what we expect might go into `config_overrides`.
3. And the common usage patterns between the two.

Reply:
That will depend on the specific plugin, but looking at Piazza, and YelloDig which both integrate via LTI. It could be:
- Consumer Key
- Consumer Secret

For Discourse it might be:
- API key
- Discourse instance url
- Enable attachments
- embed mode: lti / iframe etc

In such a case a global config could provide all the above options, and a config override could choose to override it on a course-level basis. It might not make sense for a course to override the API configuration or the discourse instance (those changes should result in a new base configuration) but there could be other settings that could be changed. 

It will be the responsibility of the plugin to disallow any changes on a course level that don't make sense. The LTI credentials for instance won't be visible to course authors at all. In that sense `course_overrides` can probably be considered more like `course_config` i.e. config options that are course-specific, however it will still shadow the global config (so that it can provide defaults) so calling it overrides is still accurate. 

I don't expect course authors to ever know that they are overriding anything that all will be handled transparently for them. Even the API that plugins ~will~ should use will not expose these separately (at least that is the intention in this PR). A plugin can request or update the course-level config and it will just know the config for that particular course without being aware of the base/parent config it's based on. This allows for some refactoring of how exactly this works behinds the scenes without breaking API.